### PR TITLE
parsercsv: Minor refactor to prefer standard algorithms and datatypes

### PR DESCRIPTION
**Pull request by tcoyvwac**
As titled. :tada:

### CODE REVIEW

File: src/library/parsercsv.cpp [hunk 1]
```diff
@@ -41,29 +41,26 @@ QList<QString> ParserCsv::parseAllLocations(const QString& playlistFile) {
 
         QList<QList<QString>> tokens = tokenize(bytes, ',');
 
-        // detect Location column
-        int locationColumnIndex = -1;
+        const auto detect_location_column = [&](const auto& tokens_list, auto predicate) {
+            const auto it = std::find_if(std::begin(tokens_list), std::end(tokens_list), predicate);
+            return (it != std::end(tokens_list))
+                    ? std::distance(std::begin(tokens_list), it)
+                    : std::optional<size_t>{std::nullopt};
```
Comment (Swiftb0y) [inline src/library/parsercsv.cpp:48]:
```suggestion
        const auto detect_location_column = [&](const auto& tokens_list, auto predicate) -> std::optional<std::size_t> {
            const auto it = std::find_if(std::begin(tokens_list), std::end(tokens_list), predicate);
            return (it != std::end(tokens_list))
                    ? std::distance(std::begin(tokens_list), it)
                    : std::nullopt;
```

File: src/library/parsercsv.cpp [hunk 2]
```diff
@@ -41,29 +41,26 @@ QList<QString> ParserCsv::parseAllLocations(const QString& playlistFile) {
...
-                }
+                locationColumnIndex = detect_location_column(tokens[1],
+                        [&](auto i) { return i.contains(QDir::separator()); });
             }
-            if (locationColumnIndex >= 0) {
+            if (locationColumnIndex.has_value()) {
                 for (int row = 1; row < tokens.size(); ++row) {
                     if (locationColumnIndex < tokens[row].size()) {
-                        locations.append(tokens[row][locationColumnIndex]);
+                        locations.append(tokens[row][*locationColumnIndex]);
```
Comment (Swiftb0y) [inline src/library/parsercsv.cpp:63]:
```suggestion
                        locations.append(tokens[row][static_cast<int>(*locationColumnIndex)]);
```

File: src/library/parsercsv.cpp [hunk 3]
```diff
@@ -41,29 +41,26 @@ QList<QString> ParserCsv::parseAllLocations(const QString& playlistFile) {
...
-                    }
-                }
+                locationColumnIndex = detect_location_column(tokens[1],
+                        [&](auto i) { return i.contains(QDir::separator()); });
             }
-            if (locationColumnIndex >= 0) {
+            if (locationColumnIndex.has_value()) {
                 for (int row = 1; row < tokens.size(); ++row) {
                     if (locationColumnIndex < tokens[row].size()) {
-                        locations.append(tokens[row][locationColumnIndex]);
```
Comment (Swiftb0y) [inline src/library/parsercsv.cpp:66]: would you mind making this range-based for as well?

**Review by Swiftb0y — COMMENTED** [review]
thanks. a couple minor comments

File: src/library/parsercsv.cpp (hunk 1 above)
Comment (tcoyvwac) [reply src/library/parsercsv.cpp:48]: Hi @Swiftb0y, sadly compilers cannot work with your suggestion. :crying_cat_face: 
The compilers need to guess the datatype to be assigned to the `return` keyword by the point of the  `?`-keyword in the ternary operator. 
`std::nullopt` is too vague of a (partially-specialised) datatype for the compiler to "strong-type"; the trailing-return datatype-hint does nothing in this regard sadly.

---
How about this code as a middle-ground?
```cpp
const auto detect_location_column = [&](const auto& tokens_list, auto predicate) {
    const auto it = std::find_if(std::begin(tokens_list), std::end(tokens_list), predicate);
    return (it != std::end(tokens_list))
            ? std::distance(std::begin(tokens_list), it)
            : std::optional<size_t>{};
};
```

Comment (Swiftb0y) [reply src/library/parsercsv.cpp:48]: How about this? I primarily just care about the trailing return type for doc purposes. Also in C++ its preferred to use the prefixed `std::size_t` instead of C's `size_t` AFAIK.
```suggestion
        const auto detect_location_column = [&](const auto& tokens_list, auto predicate) -> std::optional<std::size_t>{
            const auto it = std::find_if(std::begin(tokens_list), std::end(tokens_list), predicate);
            return (it != std::end(tokens_list))
                    ? std::distance(std::begin(tokens_list), it)
                    : std::optional<std::size_t>{};
```

File: src/library/parsercsv.cpp [hunk 4]
```diff
@@ -41,31 +41,33 @@ QList<QString> ParserCsv::parseAllLocations(const QString& playlistFile) {
...
-                }
+            if (locationColumnIndex.has_value()) {
+                std::for_each(std::next(std::begin(tokens)),
+                        std::end(tokens),
+                        [&](const auto& token) {
+                            if (locationColumnIndex < token.size()) {
+                                locations.append(token[static_cast<int>(
+                                        *locationColumnIndex)]);
+                            }
+                        });
```
Comment (Swiftb0y) [inline src/library/parsercsv.cpp:70]: Mhmm. the code has just gotten more complex. I didn't anticipate that. My fault. Would you mind reverting the change again. My apology.

Comment (tcoyvwac) [reply src/library/parsercsv.cpp:70]: Yeah :crying_cat_face: , the Mixxx project does not use the `range-v3` library yet (hint-hint! :wink:)... Then the code would be as simple as:
```cpp
for(const auto& token: ranges::views::tail(tokens)) {
  if (locationColumnIndex < token.size()) {
      locations.append(token[static_cast<int>(
              *locationColumnIndex)]);
  }
}
```
and many more nice things... :smiling_face_with_tear: _(`range-v3` stuff will arrive in `C++23`)_
I agree with you jointly that a range-based approach should be the default writing-style / default way to go.

Will revert back to the previous raw-for-loop form.

Comment (Swiftb0y) [reply src/library/parsercsv.cpp]: Yeah... but expect to wait a bit for that... It was already a pain to switch to C++20...
You're welcome to add the `ranges-v3` library as a sort of polyfill to mixxx if you'd like. Do you have any good resources to recommend to learn about ranges? I don't know any mixxx maintainer that knows how to use them (me included).

Comment (tcoyvwac) [reply src/library/parsercsv.cpp]: Interesting question @Swiftb0y, humm, it really depends on one's C++ knowledge.
The tl;dr form is:
1. Understanding Standard Library algorithms and why they are important (C++03 minimum).
2. Understanding the concept & reasoning behind [boost::range](https://www.boost.org/doc/libs/release/libs/range/) (Basically `range-v2`).
  (_Why is this `ranges-v2`? From the [History section of boost::range](https://www.boost.org/doc/libs/1_80_0/libs/range/doc/html/range/history_ack.html):_)
  `This version introduced Range Adaptors and Range Algorithms. This version 2 is the result of a merge of all of the RangeEx features into Boost.Range.`

  * 2.1 [boost::range - Introduction](https://www.boost.org/doc/libs/1_80_0/libs/range/doc/html/range/introduction.html)
  * 2.2 [boost::range - Range Adapter - Introduction & Motivation](https://www.boost.org/doc/libs/1_80_0/libs/range/doc/html/range/reference/adaptors/introduction.html)
  * 2.3 [boost::range - Range Algorithm - Introduction & Motivation](https://www.boost.org/doc/libs/1_80_0/libs/range/doc/html/range/reference/algorithms/introduction.html)
  * 2.4 Bonus Algorithms Resource: - [Christian Aichinger - Boost Range For Humans](https://www.caichinger.com/boost-range/index.html)
3. Understanding about [`range-v3`](https://github.com/ericniebler/range-v3)
  * [range-v3 - Quick Start](https://ericniebler.github.io/range-v3/index.html#tutorial-quick-start)
  * [Code Project - An Introduction to the Range-v3 Library](https://www.codeproject.com/Articles/5276756/An-Introduction-to-the-Range-v3-Library)

---
(Sadly,) This is a quick writeup, so thats all for now! Hope this helps and guides on your ranges-concept journey, ye? :smile_cat:

Comment (Swiftb0y) [reply src/library/parsercsv.cpp]: Thanks a bunch. I'll definitely check those links out when I find the time.

**Review by Swiftb0y — APPROVED** [review]
lgtm. thanks

**Comment by tcoyvwac** [comment]
Thanks @Swiftb0y for the... swift PR review and merge! Very grateful ye? :smile_cat: :tada:
